Description: Updating code to utilize exceptions instead of using warn
 This update is used for the Patch extension and the Strict extension
Forwarded: no
Author: Carl Fürstenberg <azatoth@gmail.com>


Index: libparse-debcontrol-perl/lib/Parse/DebControl/Error.pm
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ libparse-debcontrol-perl/lib/Parse/DebControl/Error.pm	2009-12-01 22:55:59.000000000 +0100
@@ -0,0 +1,104 @@
+use strict;
+use warnings;
+
+package Parse::DebControl::Error;
+=pod
+
+=head1 NAME
+Parse::DebControl::Error - Exception classes for Parse::DebControl
+
+=head1 SYNOPSIS
+
+    use Parse::DebControl::Error;
+
+    throw Parse::DebControl::Error();
+
+    throw Parse::DebControl::Error::Parse( "reason for exception" );
+    throw Parse::DebControl::Error::Parse( "reason for exception", $line_number_of_data );
+    throw Parse::DebControl::Error::Parse( "reason for exception", $line_number_of_data, $context_line );
+
+    throw Parse::DebControl::Error::IO( "information regarding the error" );
+
+=head1 COPYRIGHT
+
+Parse::DebControl is copyright 2003,2004 Jay Bonci E<lt>jaybonci@cpan.orgE<gt>.
+
+Parse::DebControl::Error is copyright 2009 Carl Fürstenberg E<lt>azatoth@gmail.comE<gt>.
+
+This program is free software; you can redistribute it and/or modify it under
+the same terms as Perl itself.
+
+=cut
+use base 'Error';
+our $VERSION = '0.1';
+sub new
+{
+    my $self = shift;
+    local $Error::Depth = $Error::Depth + 1;
+
+    $self->SUPER::new(@_);
+}
+
+package Parse::DebControl::Error::Parse;
+
+use base 'Parse::DebControl::Error';
+our $VERSION = '0.1';
+
+sub new
+{
+    my $self = shift;
+    my $text = "".shift;
+    my @args = ();
+
+    my $line = shift;
+    my $context = shift;
+
+    push(@args, '-context', $context) if defined($context);
+    push(@args, '-line', $line) if defined($line);
+
+    local $Error::Depth = $Error::Depth + 1;
+
+    $self->SUPER::new(-text => $text, @args);
+}
+
+sub stringify {
+    my $self = shift;
+    my $text;
+    if( $self->context ) {
+        $text = sprintf("Parse error: %s at line %d of data (\"%s\").\n",  $self->SUPER::stringify, $self->line, $self->context);
+    } elsif( $self->line ) {
+        $text = sprintf("Parse error: %s at line %d of data.\n",  $self->SUPER::stringify, $self->line);
+    } else {
+        $text = sprintf("Parse error: %s.\n", $self->SUPER::stringify);
+    }
+    $text;
+}
+
+sub context {
+    my $self = shift;
+    exists $self->{'-context'} ? $self->{'-context'} : undef;
+}
+
+package Parse::DebControl::Error::IO;
+
+use base 'Parse::DebControl::Error';
+our $VERSION = '0.1';
+sub new
+{
+    my $self = shift;
+    my $text = "".shift;
+    my @args = ();
+
+    local $Error::Depth = $Error::Depth + 1;
+
+    $self->SUPER::new(-text => $text, @args);
+}
+
+sub stringify {
+    my $self = shift;
+    my $text;
+    $text = sprintf("IO error: %s.\n", $self->SUPER::stringify );
+    $text;
+}
+
+1;
Index: libparse-debcontrol-perl/lib/Parse/DebControl.pm
===================================================================
--- libparse-debcontrol-perl.orig/lib/Parse/DebControl.pm	2009-12-01 22:51:46.000000000 +0100
+++ libparse-debcontrol-perl/lib/Parse/DebControl.pm	2009-12-01 23:01:10.000000000 +0100
@@ -13,6 +13,7 @@
 use IO::Scalar;
 use Compress::Zlib;
 use LWP::UserAgent;
+use Parse::DebControl::Error;
 
 use vars qw($VERSION);
 $VERSION = '2.005';
@@ -33,15 +34,13 @@
 	my ($this, $filename, $options) = @_;
 	unless($filename)
 	{
-		$this->_dowarn("parse_file failed because no filename parameter was given");
-		return;
+		throw Parse::DebControl::Error::IO("parse_file failed because no filename parameter was given");
 	}	
 
 	my $fh;
 	unless(open($fh,"$filename"))
 	{
-		$this->_dowarn("parse_file failed because $filename could not be opened for reading");
-		return;
+		throw Parse::DebControl::Error::IO("parse_file failed because $filename could not be opened for reading");
 	}
 	
 	return $this->_parseDataHandle($fh, $options);
@@ -52,16 +51,14 @@
 
 	unless($data)
 	{
-		$this->_dowarn("parse_mem failed because no data was given");
-		return;
+		throw Parse::DebControl::Error::IO("parse_mem failed because no data was given");
 	}
 
 	my $IOS = new IO::Scalar \$data;
 
 	unless($IOS)
 	{
-		$this->_dowarn("parse_mem failed because IO::Scalar creation failed.");
-		return;
+		throw Parse::DebControl::Error::IO("parse_mem failed because IO::Scalar creation failed.");
 	}
 
 	return $this->_parseDataHandle($IOS, $options);
@@ -73,8 +70,7 @@
 
 	unless($url)
 	{
-		$this->_dowarn("No url given, thus no data to parse");
-		return;
+		throw Parse::DebControl::Error::IO("No url given, thus no data to parse");
 	}
 
 	my $ua = LWP::UserAgent->new;
@@ -83,8 +79,7 @@
 
 	unless($request)
 	{
-		$this->_dowarn("Failed to instantiate HTTP Request object");
-		return;
+		throw Parse::DebControl::Error::IO("Failed to instantiate HTTP Request object");
 	}
 
 	my $response = $ua->request($request);
@@ -92,8 +87,7 @@
 	if ($response->is_success) {
 		return $this->parse_mem($response->content(), $options);
 	} else {
-		$this->_dowarn("Failed to fetch $url from the web");
-		return;
+		throw Parse::DebControl::Error::IO("Failed to fetch $url from the web");
 	}
 }
 
@@ -102,22 +96,19 @@
 
 	unless($filenameorhandle)
 	{
-		$this->_dowarn("write_file failed because no filename or filehandle was given");
-		return;
+		throw Parse::DebControl::Error::IO("write_file failed because no filename or filehandle was given");
 	}
 
 	unless($dataorarrayref)
 	{
-		$this->_dowarn("write_file failed because no data was given");
-		return;
+		throw Parse::DebControl::Error::IO("write_file failed because no data was given");
 	}
 
 	my $handle = $this->_getValidHandle($filenameorhandle, $options);
 
 	unless($handle)
 	{
-		$this->_dowarn("write_file failed because we couldn't negotiate a valid handle");
-		return;
+		throw Parse::DebControl::Error::IO("write_file failed because we couldn't negotiate a valid handle");
 	}
 
 	my $string = $this->write_mem($dataorarrayref, $options);
@@ -134,8 +125,7 @@
 
 	unless($dataorarrayref)
 	{
-		$this->_dowarn("write_mem failed because no data was given");
-		return;
+		throw Parse::DebControl::Error::IO("write_mem failed because no data was given");
 	}
 
 	my $arrayref = $this->_makeArrayref($dataorarrayref);
@@ -165,8 +155,7 @@
 	{
 		unless($filenameorhandle->opened())
 		{
-			$this->_dowarn("Can't get a valid filehandle to write to, because that is closed");
-			return;
+			throw Parse::DebControl::Error::IO("Can't get a valid filehandle to write to, because that is closed");
 		}
 
 		return $filenameorhandle;
@@ -180,8 +169,7 @@
 
 		unless(open $handle,"$openmode$filenameorhandle")
 		{
-			$this->_dowarn("Couldn't open file: $openmode$filenameorhandle for writing");
-			return;
+			throw Parse::DebControl::Error::IO("Couldn't open file: $openmode$filenameorhandle for writing");
 		}
 
 		return $handle;
@@ -248,8 +236,7 @@
 
 	unless($handle)
 	{
-		$this->_dowarn("_parseDataHandle failed because no handle was given. This is likely a bug in the module");
-		return;
+		throw Parse::DebControl::Error("_parseDataHandle failed because no handle was given. This is likely a bug in the module");
 	}
 
 	if($options->{tryGzip})
@@ -309,8 +296,7 @@
 
 				$lastfield = $key;
 			}else{
-				$this->_dowarn("Parse error on line $linenum of data; invalid key/value stanza");
-				return $structs;
+                throw Parse::DebControl::Error::Parse('invalid key/value stansa', $linenum, $line);
 			}
 
 		}elsif($line =~ /^([\t\s])(.*)/)
@@ -319,8 +305,7 @@
 
 			unless($lastfield)
 			{
-				$this->_dowarn("Parse error on line $linenum of data; indented entry without previous line");
-				return $structs;
+                throw Parse::DebControl::Error::Parse("indented entry without previous line", $linenum);
 			}
 			if($options->{verbMultiLine}){
 				$data->{$lastfield}.="\n$1$2";
@@ -343,8 +328,7 @@
 			$data = $this->_getReadyHash($options);
 			$lastfield = "";
 		}else{
-			$this->_dowarn("Parse error on line $linenum of data; unidentified line structure");
-			return $structs;
+            throw Parse::DebControl::Error::Parse("unidentified line structure", $linenum, $line);
 		}
 
 	}
@@ -379,8 +363,7 @@
 		eval("use Tie::IxHash");
 		if($@)
 		{
-			$this->_dowarn("Can't use Tie::IxHash. You need to install it to have this functionality");
-			return;
+			throw Parse::DebControl::Error("Can't use Tie::IxHash. You need to install it to have this functionality");
 		}
 		tie(%$data, "Tie::IxHash");
 		return $data;
@@ -389,19 +372,6 @@
 	return {};
 }
 
-sub _dowarn
-{
-        my ($this, $warning) = @_;
-
-        if($this->{_verbose})
-        {
-                warn "DEBUG: $warning";
-        }
-
-        return;
-}
-
-
 1;
 
 __END__
